### PR TITLE
Leaderboard v4

### PR DESCRIPTION
**Pull request by phelios**
- Add date range into title printing
- Allow date mocking through env
- Allow users to specify a date when requesting for weekly and monthly leaderboards

### CODE REVIEW

**Comment by phelios** [comment]
wooh.. first PR with green build on first try! Made sure I used double quotes too ;)

**Comment by michael-robbins** [comment]
Deserves an instant approve just for that alone lol

**Review by michael-robbins — CHANGES_REQUESTED** [review]

File: emojirades/analytics/leaderboard.py [hunk 1]
```diff
@@ -127,18 +129,15 @@ def calculate_score(history):
             for u in sorted(leaderboard, key=leaderboard.get, reverse=True)
         ]
 
-    def get_by_range(self, of_date: pendulum.DateTime, unit: str):
-        start_time = of_date.start_of(unit).timestamp()
-        end_time = of_date.end_of(unit).timestamp()
+    def get_by_range(self, of_date: pendulum.DateTime, time_unit):
```
[the remark this answers is not in the document]
Comment (phelios) [reply emojirades/analytics/leaderboard.py:132]: yeah man, seriously the "optional typing" is throwing me off coz I'm lazy lol. It's now typed to the TimeRange constants, which should be enums, so I'll change that.

File: emojirades/commands/scorekeeper_commands/leaderboard_command.py [hunk 2]
```diff
@@ -26,23 +31,30 @@ def __init__(self, *args, **kwargs):
...
             self.logger.debug(f"Getting a {self.time_unit} leaderboard")
             history = self.scorekeeper.raw_history(self.args["channel"])
-            of_date = pendulum.now("Australia/Melbourne")
-            lb = LeaderBoard(history)
 
-            if self.time_unit == TimeRange.WEEKLY:
-                leaderboard = lb.get_week(of_date)
-            elif self.time_unit == TimeRange.MONTHLY:
-                leaderboard = lb.get_month(of_date)
+            mock_date = os.environ.get("EMOJIRADE_MOCK_DATE")
```
[the remark this answers is not in the document]
Comment (phelios) [reply emojirades/commands/scorekeeper_commands/leaderboard_command.py:47]: heheh.. can it? 😬  Behave is using it to get a consistent leaderboard output.

File: emojirades/printers/leaderboard_printer.py [hunk 3]
```diff
@@ -17,7 +31,19 @@ def print(self):
...
 
-        lines = ["```", f" :: {self.time_unit.title()} leaderboard ::", ""]
+        title = []
+        title.append("")
+        title.append("::")
+        title.append(f"{self.time_unit.title()}")
+        title.append("leaderboard")
+        if self.print_date_range():
+            title.append(self.print_date_range())
+        title.append("::")
```
[the remark this answers is not in the document]
Comment (phelios) [reply emojirades/printers/leaderboard_printer.py:41]: I though it's easier to look at this way. Also coz I'll have to handle the extra space in ALL TIME leaderboard where there's no date range.

File: emojirades/commands/scorekeeper_commands/leaderboard_command.py [hunk 4]
```diff
@@ -26,23 +31,30 @@ def __init__(self, *args, **kwargs):
...
+            if self.on_date:
+                of_date = pendulum.from_format(self.on_date, "YYYYMMDD", tz=self.TZ)
+                self.logger.info(f"User requested date to: {of_date}")
+            elif mock_date:
+                # Mockable date
+                of_date = pendulum.from_format(mock_date, "YYYYMMDD", tz=self.TZ)
+                self.logger.info(f"Mocking date to: {of_date}")
```
[the remark this answers is not in the document]
Comment (phelios) [reply emojirades/commands/scorekeeper_commands/leaderboard_command.py:54]: hmm.. there's an of_date that's set to `now()` at the top of the method. I see why this is not very clear, but the alternative is moving it down here as else statement but would look weird too for the printer coz it's not part of the if-else. Any better idea?

[the remark this answers is not in the document]
Comment (michael-robbins) [reply emojirades/commands/scorekeeper_commands/leaderboard_command.py:54]: Sorry missed that, yep that's fine!

**Review by michael-robbins — APPROVED** [review]

**Review by michael-robbins — APPROVED** [review]

**Pull request merged**